packages/core/src/components/forms/fileUpload.tsx

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the terms of the LICENSE file distributed with this project.
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import { Utils } from "../../common";
+import * as Classes from "../../common/classes";
+import { IProps } from "../../common/props";
+
+export interface IFileUploadProps extends React.HTMLProps<HTMLLabelElement>, IProps {
+    /**
+     * Whether the file upload is non-interactive.
+     * Setting this to `true` will automatically disable the child input too.
+     */
+    disabled?: boolean;
+
+    /**
+     * Whether the file upload should take up the full width of its container.
+     */
+    fill?: boolean;
+
+    /**
+     * The props to pass to the child input.
+     * `disabled` will be ignored in favor of the top-level prop.
+     * `type` will be ignored, because the input _must_ be `type="file"`.
+     * Pass `onChange` here to be notified when the user uploads a file.
+     */
+    inputProps?: React.HTMLProps<HTMLInputElement>;
+
+    /**
+     * Whether the file upload should appear with large styling.
+     */
+    large?: boolean;
+
+    /**
+     * Callback invoked on `input` `change` events.
+     *
+     * This callback is offered as a convenience; it is equivalent to passing
+     * `onChange` to `inputProps`.
+     *
+     * __Note:__ If you pass `onChange` as a top-level prop, it will be passed
+     * to the wrapping `label` rather than the `input`, which may not be what
+     * you expect.
```

Comment:
👌

Reply:
What does `onChange` on `<label>` even mean? Is it even valid? I really think we should consider just piping it through to the `<input>`; there doesn't seem to be a valid use case for `<label onChange={...}>` in practice.

Reply:
The component's `props` should be spreadable to the outermost node of the component. That node is a `<label>` here, so the props interface looks like this:

```
interface IFileUploadProps extends React.HTMLProps<HTMLLabelElement>, IProps {
```

Since `onChange` uses the custom type `HTMLLabelElement`, it can't be spread to the `input` (due to the type conflict). Maybe I can cast it, but I'd rather not pick and choose which HTML top-level props are spread where.

Reply:
shrug... I still don't like the API tradeoffs we're making here but we can revisit in 2.0

Reply:
oof typings biting butts again. would be nice and logical for `onChange` to apply to the input, but makes sense to spread all props to `label` element. we have precedent for `onInputChange` in other input-enveloping components.